### ISSUE TITLE
[bug] Wrong `CUDA_VERSION` for `hipsparseSpVecDescr_t` and `hipsparseDnVecDescr_t` 

### ISSUE THREAD

**Issue by emankov** [ROCmSoftwarePlatform/hipSPARSE#325]
The corresponding to `hipsparseSpVecDescr_t` and `hipsparseDnVecDescr_t`  cuSPARSE APIs `cusparseSpVecDescr_t` and `cusparseDnVecDescr_t` appeared in CUDA 10.2, not in CUDA 10.1.

https://github.com/ROCmSoftwarePlatform/hipSPARSE/blob/develop/library/include/hipsparse.h#LL8777C1-L8783C7:
```c++
/* Generic API opaque structures holding information */
#if(!defined(CUDART_VERSION) || CUDART_VERSION >= 10010)
typedef void* hipsparseSpVecDescr_t;
typedef void* hipsparseSpMatDescr_t;
typedef void* hipsparseDnVecDescr_t;
typedef void* hipsparseDnMatDescr_t;
#endif
```

**Comment by jsandham** [comment]
Hi @emankov. I am just skimming through the cusparse documentation and it appear that both cusparseSpVecDescr_t and cusparseDnVecDescr_t exist in the 10.1 version. Ill take a look at the actual cusparse shared library and see if the functions cusparseCreateDnVec are actually present or not and report back.

**Comment by jsandham** [comment]
@emankov Here is what I have determined so far:

cuSPARSE actually has three versions for 10.1, namely:
10.1
10.1 update 1
10.1 update 2

as can be seen here https://developer.nvidia.com/cuda-toolkit-archive

Now for 10.1, cusparseSpVecDescr_t and cusparseDnVecDescr_t are not defined and the shared library does not contain routines using these descriptors like cusparseCreateSpVec, cusparseCreateDnVec etc. In 10.1 update 1 and 10.1 update 2 the shared library does have them. 

Now for all three versions of 10.1, the CUDART_VERSION define is 10010. This define is what we use to determine cusparse backend compatibility. So what happens is if you have cusparse 10.1 installed then you get compile errors. If you have cusparse 10.1 update 1 or cusparse 10.1 update 2 then you should be fine.

I am in conversions about how to resolve this issue and will report back once I have a solution.

**Comment by emankov** [comment]
Starting from the `CUDA_VERSION` >= `10020`, the separate versioning of the CUDA libs including `CUSPARSE_VERSION` appeared. I use it together with `CUDA_VERSION` in some cases related to CUDA Update X differences, for instance:
```c++
#if CUDA_VERSION >= 12010 && CUSPARSE_VERSION >= 12100
  // CHECK: rocsparse_spmv_alg SPMV_SELL_ALG1 = rocsparse_spmv_alg_ell;
  cusparseSpMVAlg_t SPMV_SELL_ALG1 = CUSPARSE_SPMV_SELL_ALG1;
#endif
```
But cases with `CUDA_VERSION` < `10020` need a different solution.

**Comment by emankov** [comment]
Options other than the `#if <version>` are:
1. `version.txt` file, which was shipped till CUDA 10.2 (inclusively) 
2. `nvcc --version` with the subsequent grabbing of the X.Y.X version

**Comment by jsandham** [comment]
After some thought I think I am going to try actually read from the cusparse shared library itself and check if the symbol exists or not. 

The issue with reading from the version.txt file (or more recently the version.json) or checking nvcc is that neither one of these is strictly required for using hipsparse so I don't know if we can guarantee their existence. I don't want to have the situation where we sometimes are able to find the version.txt or nvcc on some users machines and sometimes not on others. This could create an inconsistent user experience.

Ill report back if my efforts to read from the shared library are successful.

**Comment by jsandham** [comment]
Fixed in up coming 5.7 release.
